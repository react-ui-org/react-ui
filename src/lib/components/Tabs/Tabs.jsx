import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import styles from './Tabs.scss';

export const Tabs = ({
  children,
  id,
}) => (
  <nav id={id}>
    <ul
      className={styles.list}
      id={id && `${id}__list`}
    >
      {children}
    </ul>
  </nav>
);

Tabs.defaultProps = {
  children: null,
  id: undefined,
};

Tabs.propTypes = {
  /**
   * Nested `TabsItem` elements.
   */
  children: PropTypes.node,
  /**
   * ID of the root HTML element. It also serves as base for nested element:
   * * `<ID>__list`
   */
  id: PropTypes.string,
};

export const TabsWithContext = withProviderContext(Tabs, 'Tabs');

export default TabsWithContext;
