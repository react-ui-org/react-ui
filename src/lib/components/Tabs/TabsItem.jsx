import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../provider';
import { classNames } from '../../utils/classNames';
import styles from './TabsItem.scss';

export const TabsItem = ({
  afterLabel,
  beforeLabel,
  href,
  id,
  isActive,
  label,
  onClick,
}) => (
  <li
    className={classNames(
      styles.root,
      isActive && styles.isRootActive,
    )}
    id={id}
    key={href}
  >
    <a
      href={href}
      onClick={onClick}
      className={styles.link}
      id={id && `${id}__link`}
    >
      <span className={styles.linkContent}>
        {beforeLabel}
        <span
          className={styles.label}
          id={id && `${id}__label`}
        >
          {label}
        </span>
        {afterLabel}
      </span>
    </a>
  </li>
);

TabsItem.defaultProps = {
  afterLabel: null,
  beforeLabel: null,
  id: undefined,
  isActive: false,
  onClick: null,
};

TabsItem.propTypes = {
  /**
   * Optional element shown after item's label, e.g. an icon.
   */
  afterLabel: PropTypes.node,
  /**
   * Optional element shown before item's label, e.g. an icon.
   */
  beforeLabel: PropTypes.node,
  /**
   * Item's link URL. Optionally add custom routing function to the `onClick` option to bypass
   * browser's default navigation.
   */
  href: PropTypes.string.isRequired,
  /**
   * ID of the root HTML element. It also serves as base for nested elements:
   * * `<ID>__link`
   * * `<ID>__label`
   */
  id: PropTypes.string,
  /**
   * If `true`, item is marked as active.
   */
  isActive: PropTypes.bool,
  /**
   * Item's label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Function to call on item click, e.g. custom routing function.
   */
  onClick: PropTypes.func,
};

export const TabsItemWithContext = withProviderContext(TabsItem, 'TabsItem');

export default TabsItemWithContext;
