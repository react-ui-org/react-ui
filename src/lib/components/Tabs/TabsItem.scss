// 1. Keep inactive items under Tabs' decorative bottom border and pop active ones above it.

@use "../../styles/tools/breakpoint";
@use "../../styles/tools/reset";
@use "../../styles/tools/transition";
@use "theme";

.root {
    @include transition.add(transform);

    position: relative; // 1.
    z-index: 1; // 1.
    flex: none;
    padding-top: calc(-1 * #{theme.$item-active-shift-y});
    margin-bottom: theme.$item-active-shift-y;
    transform: translateY(0);

    &:hover,
    &:focus-within {
        transform: translateY(theme.$item-hover-shift-y);
    }

    &:not(:last-child) {
        margin-right: theme.$gap-xs;
    }

    @include breakpoint.up(sm) {
        &:not(:last-child) {
            margin-right: theme.$gap-sm;
        }
    }

    @include breakpoint.up(md) {
        &:not(:last-child) {
            margin-right: theme.$gap-md;
        }
    }
}

.link {
    @include reset.link();
    @include transition.add((color, border-color, background-color, box-shadow));

    display: block;
    padding: theme.$item-padding-xs;
    font-weight: theme.$item-font-weight;
    line-height: 1;
    text-decoration: none;
    color: theme.$item-color;
    border-width: theme.$item-border-width;
    border-style: solid;
    border-color: theme.$item-border-color;
    border-top-left-radius: theme.$item-border-radius;
    border-top-right-radius: theme.$item-border-radius;
    background-color: theme.$item-background-color;
    box-shadow: theme.$item-box-shadow;

    &:hover,
    &:focus {
        font-weight: theme.$item-hover-font-weight;
        color: theme.$item-hover-color;
        border-width: theme.$item-hover-border-width;
        border-color: theme.$item-hover-border-color;
        background-color: theme.$item-hover-background-color;
        box-shadow: theme.$item-hover-box-shadow;
    }

    @include breakpoint.up(sm) {
        padding: theme.$item-padding-sm;
    }

    @include breakpoint.up(md) {
        padding: theme.$item-padding-md;
    }
}

.linkContent {
    @include transition.add(transform);

    display: flex;
    align-items: center;
    transform: translateY(0);
}

.label:not(:first-child) {
    margin-left: theme.$item-icon-gap;
}

.label:not(:last-child) {
    margin-right: theme.$item-icon-gap;
}

.isRootActive,
.isRootActive:hover,
.isRootActive:focus-within {
    z-index: 2; // 1.
    transform: translateY(theme.$item-active-shift-y);
}

.isRootActive .link {
    font-weight: theme.$item-active-font-weight;
    color: theme.$item-active-color;
    border-width: theme.$item-active-border-width;
    border-color: theme.$item-active-border-color;
    background-color: theme.$item-active-background-color;
    box-shadow: theme.$item-active-box-shadow;
}

.link:hover .linkContent,
.link:focus .linkContent {
    transform: translateY(theme.$item-hover-label-shift-y);
}

.isRootActive .link .linkContent,
.isRootActive .link:hover .linkContent,
.isRootActive .link:focus .linkContent {
    transform: translateY(theme.$item-active-label-shift-y);
}
