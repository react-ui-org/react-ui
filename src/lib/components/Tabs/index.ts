export { default as Tabs } from './Tabs';
export { default as TabsItem } from './TabsItem';
