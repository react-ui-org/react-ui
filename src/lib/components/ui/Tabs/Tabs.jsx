import PropTypes from 'prop-types';
import React from 'react';
import { withProviderContext } from '../../../provider';
import styles from './Tabs.scss';

export const Tabs = ({
  children,
  id,
}) => (
  <nav id={id}>
    <ul
      className={styles.list}
      id={id && `${id}__list`}
    >
      {children}
    </ul>
  </nav>
);

Tabs.defaultProps = {
  children: null,
  id: undefined,
};

Tabs.propTypes = {
  /**
   * Individual `TabsItem`s.
   */
  children: PropTypes.node,
  /**
   * ID of the root HTML element. It also serves as a prefix for inner list element:
   * `<ID>__list`.
   */
  id: PropTypes.string,
};

export const TabsWithContext = withProviderContext(Tabs, 'Tabs');

export default TabsWithContext;
