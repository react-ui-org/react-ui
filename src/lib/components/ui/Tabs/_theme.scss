$border-bottom-width: var(--rui-Tabs__border-bottom-width);
$border-bottom-color: var(--rui-Tabs__border-bottom-color);
$gap-xs: var(--rui-Tabs__gap--xs);
$gap-sm: var(--rui-Tabs__gap--sm);
$gap-md: var(--rui-Tabs__gap--md);
$padding-x: var(--rui-Tabs__padding-x);

$item-padding-xs: var(--rui-Tabs__item__padding--xs);
$item-padding-sm: var(--rui-Tabs__item__padding--sm);
$item-padding-md: var(--rui-Tabs__item__padding--md);
$item-font-weight: var(--rui-Tabs__item__font-weight);
$item-color: var(--rui-Tabs__item__color);
$item-border-width: var(--rui-Tabs__item__border-width);
$item-border-color: var(--rui-Tabs__item__border-color);
$item-border-radius: var(--rui-Tabs__item__border-radius);
$item-background-color: var(--rui-Tabs__item__background-color);
$item-box-shadow: var(--rui-Tabs__item__box-shadow);
$item-icon-gap: var(--rui-Tabs__item__icon__gap);

$item-hover-font-weight: var(--rui-Tabs__item--hover__font-weight);
$item-hover-color: var(--rui-Tabs__item--hover__color);
$item-hover-border-width: var(--rui-Tabs__item--hover__border-width);
$item-hover-border-color: var(--rui-Tabs__item--hover__border-color);
$item-hover-background-color: var(--rui-Tabs__item--hover__background-color);
$item-hover-box-shadow: var(--rui-Tabs__item--hover__box-shadow);
$item-hover-shift-y: var(--rui-Tabs__item--hover__shift-y);
$item-hover-label-shift-y: var(--rui-Tabs__item--hover__label__shift-y);

$item-active-font-weight: var(--rui-Tabs__item--active__font-weight);
$item-active-color: var(--rui-Tabs__item--active__color);
$item-active-border-width: var(--rui-Tabs__item--active__border-width);
$item-active-border-color: var(--rui-Tabs__item--active__border-color);
$item-active-background-color: var(--rui-Tabs__item--active__background-color);
$item-active-box-shadow: var(--rui-Tabs__item--active__box-shadow);
$item-active-shift-y: var(--rui-Tabs__item--active__shift-y);
$item-active-label-shift-y: var(--rui-Tabs__item--active__label__shift-y);
