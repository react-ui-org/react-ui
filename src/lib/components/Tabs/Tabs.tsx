import React from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import styles from './Tabs.scss';
import { TabsProps } from './Tabs.types';

export const Tabs : React.FunctionComponent<TabsProps> = ({
  children,
  id,
  ...restProps
}) => (
  <nav
    {...transferProps(restProps)}
    id={id}
  >
    <ul
      className={styles.list}
      id={id && `${id}__list`}
    >
      {children}
    </ul>
  </nav>
);

export const TabsWithGlobalProps = withGlobalProps(Tabs, 'Tabs');
export default TabsWithGlobalProps;
