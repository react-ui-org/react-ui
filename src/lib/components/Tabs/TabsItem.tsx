import React from 'react';
import { withGlobalProps } from '../../provider';
import { transferProps } from '../_helpers/transferProps';
import { classNames } from '../../utils/classNames';
import styles from './TabsItem.scss';
import { TabsItemProps } from './Tabs.types';

export const TabsItem: React.FunctionComponent<TabsItemProps> = ({
  afterLabel,
  beforeLabel,
  href,
  id,
  isActive = false,
  label,
  onClick,
  ...restProps
}) => (
  <li
    {...transferProps(restProps)}
    className={classNames(
      styles.root,
      isActive && styles.isRootActive,
    )}
    id={id}
    key={href}
  >
    <a
      className={styles.link}
      href={href}
      id={id && `${id}__link`}
      onClick={onClick}
    >
      <span className={styles.linkContent}>
        {beforeLabel}
        <span
          className={styles.label}
          id={id && `${id}__label`}
        >
          {label}
        </span>
        {afterLabel}
      </span>
    </a>
  </li>
);

export const TabsItemWithGlobalProps = withGlobalProps(TabsItem, 'TabsItem');

export default TabsItemWithGlobalProps;
