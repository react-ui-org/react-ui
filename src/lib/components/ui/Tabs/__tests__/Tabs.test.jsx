import React from 'react';
import { shallow } from 'enzyme';
import { Tabs } from '../Tabs';
import { TabsItem } from '../TabsItem';

describe('rendering', () => {
  it('renders correctly without children', () => {
    const tree = shallow(
      <Tabs />,
    );

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with children', () => {
    const tree = shallow(
      <Tabs id="my-tabs">
        <TabsItem label="item 1" href="/item-1" />
        <TabsItem label="item 2" href="/item-2" />
        <TabsItem label="item 3" href="/item-3" />
      </Tabs>,
    );

    expect(tree).toMatchSnapshot();
  });
});
