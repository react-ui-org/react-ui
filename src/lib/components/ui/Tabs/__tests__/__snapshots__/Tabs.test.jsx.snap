// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly with children 1`] = `
<nav
  id="my-tabs"
>
  <ul
    className="list"
    id="my-tabs__list"
  >
    <TabsItem
      href="/item-1"
      label="item 1"
    />
    <TabsItem
      href="/item-2"
      label="item 2"
    />
    <TabsItem
      href="/item-3"
      label="item 3"
    />
  </ul>
</nav>
`;

exports[`rendering renders correctly without children 1`] = `
<nav>
  <ul
    className="list"
  />
</nav>
`;
